pkg/istio-agent/sds-agent.go

```diff
@@ -64,13 +64,17 @@ var (
 	pluginNamesEnv             = env.RegisterStringVar(pluginNames, "", "").Get()
 	enableIngressGatewaySDSEnv = env.RegisterBoolVar(enableIngressGatewaySDS, false, "").Get()
 
-	trustDomainEnv                     = env.RegisterStringVar(trustDomain, "", "").Get()
-	secretTTLEnv                       = env.RegisterDurationVar(secretTTL, 24*time.Hour, "").Get()
-	secretRefreshGraceDurationEnv      = env.RegisterDurationVar(SecretRefreshGraceDuration, 12*time.Hour, "").Get()
-	secretRotationIntervalEnv          = env.RegisterDurationVar(SecretRotationInterval, 10*time.Minute, "").Get()
-	staledConnectionRecycleIntervalEnv = env.RegisterDurationVar(staledConnectionRecycleInterval, 5*time.Minute, "").Get()
-	initialBackoffInMilliSecEnv        = env.RegisterIntVar(InitialBackoffInMilliSec, 0, "").Get()
-	pkcs8KeysEnv                       = env.RegisterBoolVar(pkcs8Key, false, "Whether to generate PKCS#8 private keys").Get()
+	trustDomainEnv = env.RegisterStringVar(trustDomain, "", "").Get()
+	secretTTLEnv   = env.RegisterDurationVar(secretTTL, 24*time.Hour,
+		"The cert lifetime requested by istio agent").Get()
+	secretRefreshGraceDurationEnv = env.RegisterDurationVar(SecretRefreshGraceDuration, secretTTLEnv/2,
+		"The grace period for the cert rotation, by default it's half of the cert lifetime").Get()
+	secretRotationIntervalEnv = env.RegisterDurationVar(SecretRotationInterval, secretRefreshGraceDurationEnv/10,
```

Comment:
Does this apply to ingress SDS? how likely is this to cause problems if we have a high TTL? I think for sidecar, we always get a fresh cert at startup, so no issues. But for ingress its a bit different. I think the user needs to rotate and when they do we trigger an immediate update so no issue here, is that correct?

Reply:
Basically my concern was if you set your TTL to a year (for example) and the pod starts up 350 days later. It will expire in 15 days but we won't check for rotation until 35 days. Or maybe since its already passed the grace period its not an issue

Reply:
That's a good concern. Actually rotate() is not triggered for ingress SDS (reading from secret) so we are fine: https://github.com/istio/istio/blob/master/security/pkg/nodeagent/cache/secretcache.go#L500

Reply:
I think updating certs for ingress SDS is based on the watcher for K8s secret.